docs/design.md

```diff
@@ -312,22 +319,21 @@ There are several scenarios in which transactions interact:
   transaction’s commit timestamp to indicate that when/if the
   transaction does commit, it should use a timestamp *at least* as
   high. However, if the write conflict is from an SSI transaction,
-  the reader must compare priorities. If it has the higher priority,
-  it pushes the transaction’s commit timestamp, as with SI (that
+  the reader must compare priorities. If the reader has the higher priority,
+  it pushes the transaction’s commit timestamp (that
   transaction will then notice its timestamp has been pushed, and
-  restart). If it has the lower priority, it retries itself using as
+  restart). If it has the lower or same priority, it retries itself using as
   a new priority `max(new random priority, conflicting txn’s
-  priority - 1)`.
-
-- **Writer encounters uncommitted write intent with lower priority**:
-  writer aborts the conflicting transaction.
+  priority - 1)`. why max and not min?
```

Comment:
It's max and not min to prevent starvation: a transaction's priority generally increases as it retries (except in the event of a tie, which may cause the priority to decrease by 1 to avoid repeated ties with the same transaction), so it becomes more likely to win any conflicts it encounters.

The two values passed to max() are heuristics to balance the need for the retried transaction to eventually succeed with the desire to minimize future conflicts and pushes. One way the random element helps is that without it, two transactions conflicting with the same third transaction could end up with the same priority and get stuck in an endless cycle of ties. The random element ensures that this cycle will eventually break, although the higher the third transaction's priority was the longer it will take.

This could be a problem: consider a consistent scan of an entire table while write traffic is going on (like the bank example). The scan will conflict with everything, and will often lose and be restarted. Eventually its priority will be high enough that it can succeed, at which point all the other transactions that it interrupted will be allowed to run with equal priorities, all quite high. These transactions could conflict and tie many times until one of them wins a random roll to set its priority even higher. Perhaps we should be subtracting a small random number from the conflicting txn's priority instead of just 1, to break up clusters of transactions that would otherwise get the same priority.